core/src/vaults.ts

```diff
@@ -0,0 +1,378 @@
+        osmEvents = await osmContract.queryFilter(osmEventFilter, fromBlock, toBlock);
+        rpcEventFilterSegmentIndex += 1;
+        fromBlock = -EVENTS_PER_RPC_REQUEST * (rpcEventFilterSegmentIndex + 1) - EVENTS_ON_FIRST_RPC_REQUEST;
+        toBlock = -EVENTS_PER_RPC_REQUEST * rpcEventFilterSegmentIndex - EVENTS_ON_FIRST_RPC_REQUEST;
+    }
+    if (osmEvents.length === 0) {
+        return undefined;
+    }
+    const currentUnitCollateralPrice = new BigNumber(osmEvents[osmEvents.length - 1].args?.val).shiftedBy(
+        -WAD_NUMBER_OF_DIGITS
+    );
+    const lastPriceUpdateAsHex = await osmContract.zzz();
+    const lastPriceUpdateTimestampInSeconds = new BigNumber(lastPriceUpdateAsHex._hex).toNumber();
+    const priceUpdateFrequencyInSeconds = await osmContract.hop();
+
+    const priceFeedContractAddress = await osmContract.src();
+    const priceFeedContractInterface = await getContractInterfaceByName('MEDIAN_PRICE_FEED');
+    const priceFeedContract = new ethers.Contract(priceFeedContractAddress, priceFeedContractInterface, provider);
+    const feedEventsFilter = priceFeedContract.filters.LogMedianPrice(null, null);
+
+    let feedEvents: Array<Event> = [];
+    rpcEventFilterSegmentIndex = 0;
+    fromBlock = -EVENTS_ON_FIRST_RPC_REQUEST;
+    toBlock = undefined;
+    while (feedEvents.length === 0 && latestBlockNumber > rpcEventFilterSegmentIndex * EVENTS_PER_RPC_REQUEST) {
+        feedEvents = await priceFeedContract.queryFilter(feedEventsFilter, fromBlock, toBlock);
+        rpcEventFilterSegmentIndex += 1;
+        fromBlock = -EVENTS_PER_RPC_REQUEST * (rpcEventFilterSegmentIndex + 1) - EVENTS_ON_FIRST_RPC_REQUEST;
+        toBlock = -EVENTS_PER_RPC_REQUEST * rpcEventFilterSegmentIndex - EVENTS_ON_FIRST_RPC_REQUEST;
+    }
+
+    if (feedEvents.length === 0) {
+        return undefined;
+    }
+
+    const nextUnitCollateralPrice = new BigNumber(feedEvents[feedEvents.length - 1].args?.val._hex).shiftedBy(
+        -WAD_NUMBER_OF_DIGITS
+    );
+
+    return {
+        currentUnitPrice: currentUnitCollateralPrice,
+        nextUnitPrice: nextUnitCollateralPrice,
+        nextPriceChange: new Date((lastPriceUpdateTimestampInSeconds + priceUpdateFrequencyInSeconds) * 1000),
+    };
+};
+
+export const getOsmPrices = memoizee(_getOsmPrices, {
+    maxAge: CACHE_EXPIRY_MS,
+    promise: true,
+    length: 2,
+});
+
+const _fetchLiquidationRatio = async (network: string, collateralType: CollateralType) => {
+    const contract = await getContract(network, 'MCD_SPOT');
+    const collateralTypeAsHex = ethers.utils.formatBytes32String(collateralType);
+    const oracleAndLiquidationRatio = await contract.ilks(collateralTypeAsHex);
+    const liquidationRatioAsHex = oracleAndLiquidationRatio.mat;
+    const oracleAddress = oracleAndLiquidationRatio.pip;
+    const liquidationRatio = new BigNumber(liquidationRatioAsHex._hex).shiftedBy(-RAY_NUMBER_OF_DIGITS).toNumber();
```

Comment:
I am not sure it's a good idea to turn ratio that can be so precise into a number. Let's keep it as BigNumber